chore(docs): write a blog post for the Thalamus component packaging (#3232)
open-component-model/open-component-model: website/content/blog/2026-07-30-ocmifying-thalamus.md

---
title: "Creating an OCM Component for a non-trivial Application"
description: "This post has detailed explanation about how to modernize a non-trivial application as large as Thalamus. Thalamus has over 80 images it needed to localize and a deployment that involved more than 11 components."
date: 2026-07-30T10:00:00+02:00
contributors: []
tags: ["ocm", "thalamus", "component"]
draft: false
---

## The Problem

> [Thalamus](https://cobaltcore-dev.github.io/thalamus/main/) is a large project with many moving parts. At the time of

Reviewer: I would add a short explanation of the purpose of the thalamus. If someone reads this blog post and hasn't heard of it before, describing it as a 'large project with many moving parts' might sound scary. :)

Author: I mean, that's why I put the link in there. 🤔 They can just navigate there and read all about it. :)

> this writing, it has 11 components that must be bundled together so the entire stack can be deployed is one.

Reviewer 2: ```suggestion
this writing, it has 11 components that must be bundled together so the entire stack can be deployed as one.
```

> It also includes a couple of open source components for which the CRDs and helm charts also had to be included in the

Reviewer 2: ```suggestion
It also includes a couple of open source components for which the CRDs and Helm charts also had to be included in the
```

component version.

Further, there were several requirements that the deployment remains configurable because certain values must overwrite
existing defaults. And on top of that, existing defaults change from environment to environment. This means, there are

> three layers of overwrites. Global defaults that are true for _every_ environment. Then, environment specific defaults

Reviewer 2: ```suggestion
three layers of overwrites: global defaults that are true for _every_ environment. Then, environment specific defaults
```

that you don't want to overwrite every time (for example, prod and dev have different defaults). And lastly, there are
cluster specific values that you would like to have on top of all the defaulting.

This had to be coded together with a very strict deployment ordering.

## What we ended up doing

First, we created a component constructor out of the many dependencies that Thalamus has. All dependencies ended up being

> component version references like this (warning, rather long yaml file incoming):

Reviewer 2: ```suggestion
component version references like the below (warning, rather long yaml file incoming):
```

```yaml
components:
  - name: github.com/cobaltcore-dev/thalamus
    version: 0.1.0
    provider:
      name: cobaltcore-dev
    resources:
      - name: thalamus-chart
        type: helmChart
        version: 0.1.0
        input:
          type: Helm/v1
          path: helm/thalamus
          repository: charts/thalamus:0.1.0
      - name: thalamus-rgd
        type: blob
        version: 0.1.0
        input:
          type: File/v1
          path: ocm-single-component/deploy/rgd.yaml
          mediaType: application/vnd.cncf.kro.resourcegraphdefinition.v1+yaml
      - name: operator-image
        type: ociImage
        version: 0.1.0
        access:
          type: ociArtifact

>           imageReference: ghcr.io/cobaltcore-dev/thalamus:latest

Author: This is just a sample here.

        input:
          type: Helm/v1
          helmRepository: https://github.com/kubernetes-sigs/external-dns/releases/download/external-dns-helm-chart-1.19.0/external-dns-1.19.0.tgz
          repository: charts/external-dns:1.19.0
      - name: external-dns-image
        type: ociImage
        version: v0.19.0
        access:
          type: ociArtifact
          imageReference: registry.k8s.io/external-dns/external-dns:v0.19.0

  - name: github.com/cobaltcore-dev/thalamus/open-webui
    version: 15.2.0
    provider:
      name: cobaltcore-dev
    resources:
      - name: open-webui-chart
        type: helmChart
        version: 15.2.0
        input:
          type: Helm/v1
          helmRepository: https://helm.openwebui.com/open-webui-15.2.0.tgz
          repository: charts/open-webui:15.2.0
      - name: open-webui-image
        type: ociImage
        version: 0.10.2
        access:
          type: ociArtifact
          imageReference: ghcr.io/open-webui/open-webui:0.10.2
      - name: redis-image
        type: ociImage
        version: 7.4.2-alpine3.21
        access:
          type: ociArtifact
          imageReference: docker.io/library/redis:7.4.2-alpine3.21
```

> Using the above, simply run:

Reviewer 2: ```suggestion
Using the above content in a `component-constructor.yaml`, we simply ran:
```

```console
ocm add component-version \
  --constructor component-constructor.yaml \
  --repository <target-repo>/thalamus
```

> Once the component is created properly we transferred it to a new location:

Reviewer 2: ```suggestion
Once the component was created properly we transferred it to a new location:
```

```console
ocm transfer cv \
  <source-repository>//github.com/cobaltcore-dev/thalamus:0.1.0 \
  <target-repository> \
  --recursive --copy-resources --upload-as ociArtifact
```

> The `--upload-as ociArtifact` here was important, because all the images and resources where OCI artifacts and [oras](https://oras.land/)

Reviewer 2: ```suggestion
The `--upload-as ociArtifact` here was important, because all the images and resources are OCI artifacts and [oras](https://oras.land/)
```

(OCM's internal OCI handling SDK) is capable of Streaming OCI -> OCI content efficiently. Which was really much needed
considering how many images and how many gigabytes of data we had to transfer.

> Then, we created the bootstrapping architecture that is similar to what the sovereign scenario has.

Reviewer 2: This is the first time the "sovereign scenario" is mentioned. Maybe you already use the link here you use in line 403.

Author: I can link to the README of it. It's not worth explaining it twice I recon. :)

The generated RGD from all of these was quite enormous. It's +1300 lines of replacements, localizations and installations.

However, using ocm-k8s-toolkit and only four Kubernetes objects:
```yaml
apiVersion: delivery.ocm.software/v1alpha1
kind: Repository
metadata:
  name: thalamus-repository
  namespace: thalamus
spec:
  repositorySpec:
    baseUrl: <target-repo>/thalamus-transferred
    type: OCIRegistry
  interval: 10m
---
apiVersion: delivery.ocm.software/v1alpha1
kind: Component
metadata:
  name: thalamus-component
  namespace: thalamus
spec:
  component: github.com/cobaltcore-dev/thalamus
  repositoryRef:
    name: thalamus-repository
  semver: "0.1.0"
  interval: 1m
---
apiVersion: delivery.ocm.software/v1alpha1
kind: Resource
metadata:
  name: thalamus-resource-rgd
  namespace: thalamus
spec:
  componentRef:
    name: thalamus-component
  resource:
    byReference:
      resource:
        name: thalamus-rgd
---
apiVersion: delivery.ocm.software/v1alpha1
kind: Deployer
metadata:
  name: thalamus-deployer
  namespace: thalamus
spec:
  resourceRef:
    name: thalamus-resource-rgd
    namespace: thalamus
```

> ... we were able to deploy the entire application architecture without problems with all the image referenced

Reviewer 2: ```suggestion
... we were able to deploy the entire application architecture without problems with all the image references
```

> pointing to the new locations (<target-repository>).

Reviewer 3: I believe md renderer is triggered by `<` and `>`

<img width="1059" height="83" alt="Image" src="https://github.com/user-attachments/assets/d6765148-7e09-47e2-b5ac-58fee18cc6bb" />

Author: Fixed. :)

### LLM usage and Skill

> Because of some prior work we did in open-component-model v2, it was quite easy to use a popular LLM (Claude, for example)

Reviewer 2: "Because of some prior work we did in open-component-model v2" is quite vague. I would spend some explanation about the why and also remove the v2.

Author: It's explained down below with the headings, Schema and Samples... But I'll make it explicit.

Author: Added explicit wording. :)

to almost one-shot a complete component version for the entire Thalamus project.

#### Schema

> The component constructor YAML file contains a Schema file located here: [component constructor schema](https://ocm.software/docs/reference/component-constructor/).

Reviewer 2: ```suggestion
The component constructor YAML file follows the schema depicted [here](https://ocm.software/docs/reference/component-constructor/).
```

> This can be provided to the LLM as a basis of generating a valid component version. It doesn't really require any

Reviewer 2: ```suggestion
This can be provided to the LLM as a basis for generating a valid component version. It doesn't really require any
```

particular skills on the LLM's side. Just link the schema it needs to read and understand.

Link in the file and tell your favorite agent the following:
> Analyze this JSON schema save its structure in your memory. Any further files names component-constructor.yaml MUST
> use this schema. Anytime I tell you to "create a component constructor" use this schema to generate the corresponding
> YAML file.

> Claude already has in-built JSON schema skill.

Reviewer 2: ```suggestion
Claude Code already has a built-in JSON schema skill.
```

Author: Claude in general has it, not just claude code.

#### Samples

We created ample of examples that create and consume complex component constructors and RGDs. One of the most complicated
one is the sovereign scenario we put together located here: [Sovereign Conformance Test](https://github.com/open-component-model/open-component-model/tree/main/conformance/scenarios/sovereign).

> Before giving it the samples however, you can also fine-tune it's understanding about Kro. Link the RGD schema definition

Reviewer 2: ```suggestion
Before giving the LLM the samples however, you can also fine-tune it's understanding about Kro. Link the RGD schema definition
```

from here: [Kro RGD JSON Schema](https://raw.githubusercontent.com/kubernetes-sigs/kro/refs/heads/main/helm/crds/kro.run_resourcegraphdefinitions.yaml). It's better to link the YAML file rather

> than point it to a website like the rendered schema on Kro's page here: [Rendered RGD](https://kro.run/api/crds/resourcegraphdefinition/). This is

Reviewer 2: ```suggestion
than point the LLM to a website like the rendered schema on Kro's page here: [Rendered RGD](https://kro.run/api/crds/resourcegraphdefinition/). This is
```

> much harder to parse and takes up too much context to process and then keep "in mind". The more specific the better.

Reviewer 2: ```suggestion
much harder to parse and takes up too much context to process and then keep "in mind". The more specific, the better.
```

> Once it understands Kro, now is a good time to point it at the folder containing the sovereign scenario then make the

Reviewer 2: ```suggestion
Once it understands Kro, now is a good time to point it at the folder containing the sovereign scenario, then make the
```

> Once it understands Kro, now is a good time to point it at the folder containing the sovereign scenario then make the

Reviewer 2: ```suggestion
Once the LLM understands Kro, now is a good time to point it at the folder containing the sovereign scenario then make the
```

following prompt:

> Read the example in this folder. It contains a complex example for structuring and creating component versions. The example
> is a multi-component product localized via Kro and deployed by Flux. Pay special attention to the localization where
> Kro CEL expressions replace values inside Helm values.yaml files.

> Once it understands it, we can move on to the next step.

Reviewer 2: ```suggestion
Once the model understands it, we can move on to the next step.
```

Author: Why is it a model here but `the LLM` all the time above? :D

#### The Next Step

> With this done, now you can begin ocmifying your application. Give it the following prompt:

Reviewer 2: ```suggestion
With this done, you can now begin ocmifying your application. Give the model the following prompt:
```

Author: Here model too and then down below LLM :D

> Analyze the deployment and packaging process of this application. (You can be more specific here if you have a helm

> > chart just point it at it.). Now, take this information and create a component constructor file that packages all of

Reviewer 2: ```suggestion
> chart just point the LLM at it.). Now, take this information and create a component constructor file that packages all of
```

> the application components as component references.

> Notice that we aren't doing everything in one step. That just overloads the context. So once the component constructor

Reviewer 2: ```suggestion
Notice that we aren't doing everything in one step. That will just overload the context. So once the component constructor
```

> exists and it looks good, now you can move on to the deployment part.

Reviewer 2: ```suggestion
exists and looks good, you can move on to the deployment part.
```

> > Create a bootstrapping structure for the application component version. The RGD should localize any and all image references

Reviewer 2: ```suggestion
> Create a bootstrapping structure for the application component version. The RGD should localize all image references
```

> found in the deployment helm charts. (If you have kustomize or something else, replace this with that.)

This should already create you a pretty close-by component version and bootstrapping architecture of your application.

## Another choice

Instead of one RGD that rules them all, we also tried creating several RGDs for separate components and then use a root
RGD and [RGD Chaining](https://kro.run/docs/building-abstractions/rgd-chaining/) in Kro. We also constructed separate

> component constructors for each individual components. The idea here was that separate components could be exchanged for

Reviewer 2: ```suggestion
component constructors for each individual component. The idea was that separate components could be exchanged for
```

organization specific ones. For example, if you had your own gpu-operator implementation that used a specific, platform

> dependent implementation that you should be able to configure the deployment to use that instead.

Reviewer 2: ```suggestion
dependent implementation then you should be able to configure the deployment to use that instead.
```

> Basically creating a plug-and-play style architecture for the components where a component can be exchanged for a different

Reviewer 2: ```suggestion
We basically created a plug-and-play style architecture for the components where a component can be exchanged for a different
```

> We basically created a plug-and-play style architecture for the components where a component can be exchanged for a different

Reviewer 4: ```suggestion
We created a plug-and-play style architecture for the components where a component can be exchanged for a different
```

Author: Fixed this by hand on another push.

one and the whole would still be valid and would work as intended.

This was not feasible in the end however, because it introduced a silent API that was not easy to discover and understand by
those who use the component. How would you know what the parts are that you can use _instead_ of the default configured
one? How would you be able to discover new components provided by different component authors? How would you even know

> that you can mix-and-match?

Reviewer 2: ```suggestion
you can mix and match?
```

> And further, if you do change a component and an inevitable failures occurs, Support wouldn't be able to help you because

Reviewer 2: ```suggestion
And further, if you do change a component and an inevitable failures occurs, a support team wouldn't be able to help you because
```

you are using a non-standard exchange for a component.

And lastly, deployment ordering and bootstrapping would be difficult to set up. RGD chaining gives you _some_ control
over ordering but having everything in one place makes `dependson` trivial to set up. The real pain would be bootstrapping

> _every_ separate component. The bootstrap suddenly was 11 x 3 (eleven times the component, resource, deployer) to deploy

Reviewer 2: ```suggestion
_every_ separate component. The bootstrap suddenly was 11 x 3 (eleven times the component, resource, deployer CRs) to deploy
```
